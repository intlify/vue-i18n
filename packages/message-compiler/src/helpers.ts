export const enum HelperNameMap {
  LIST = 'list',
  NAMED = 'named',
  PLURAL = 'plural',
  LINKED = 'linked',
  MESSAGE = 'message',
  TYPE = 'type',
  INTERPOLATE = 'interpolate',
  NORMALIZE = 'normalize'
}
