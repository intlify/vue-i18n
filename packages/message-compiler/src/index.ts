export * from './options'
export * from './errors'
export * from './helpers'
export * from './location'
export * from './parser'
export * from './compiler'
