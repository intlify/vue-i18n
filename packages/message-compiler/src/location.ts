// The source range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source?: string
}

export const LocationStub: SourceLocation = {
  start: { line: 1, column: 1, offset: 0 },
  end: { line: 1, column: 1, offset: 0 }
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

export function createPosition(
  line: number,
  column: number,
  offset: number
): Position {
  return { line, column, offset } as Position
}

export function createLocation(
  start: Position,
  end: Position,
  source?: string
): SourceLocation {
  const loc = { start, end } as SourceLocation
  if (source != null) {
    loc.source = source
  }
  return loc
}
