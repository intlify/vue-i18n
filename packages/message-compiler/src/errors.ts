import { format } from '@intlify/shared'
import { SourceLocation } from './location'

export type CompileDomain =
  | 'tokenizer'
  | 'parser'
  | 'generator'
  | 'transformer'
  | 'compiler'

export interface CompileError extends SyntaxError {
  code: number
  domain?: CompileDomain
  location?: SourceLocation
}

export interface CreateCompileErrorOptions {
  domain?: CompileDomain
  messages?: { [code: number]: string }
  args?: unknown[]
}

export const enum CompileErrorCodes {
  // tokenizer error codes
  EXPECTED_TOKEN,
  INVALID_TOKEN_IN_PLACEHOLDER,
  UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER,
  UNKNOWN_ESCAPE_SEQUENCE,
  INVALID_UNICODE_ESCAPE_SEQUENCE,
  UNBALANCED_CLOSING_BRACE,
  UNTERMINATED_CLOSING_BRACE,
  EMPTY_PLACEHOLDER,
  NOT_ALLOW_NEST_PLACEHOLDER,
  INVALID_LINKED_FORMAT,

  // parser error codes
  MUST_HAVE_MESSAGES_IN_PLURAL,
  UNEXPECTED_EMPTY_LINKED_MODIFIER,
  UNEXPECTED_EMPTY_LINKED_KEY,
  UNEXPECTED_LEXICAL_ANALYSIS,

  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  __EXTEND_POINT__
}

/** @internal */
export const errorMessages: { [code: number]: string } = {
  // tokenizer error messages
  [CompileErrorCodes.EXPECTED_TOKEN]: `Expected token: '{0}'`,
  [CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER]: `Invalid token in placeholder: '{0}'`,
  [CompileErrorCodes.UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER]: `Unterminated single quote in placeholder`,
  [CompileErrorCodes.UNKNOWN_ESCAPE_SEQUENCE]: `Unknown escape sequence: \\{0}`,
  [CompileErrorCodes.INVALID_UNICODE_ESCAPE_SEQUENCE]: `Invalid unicode escape sequence: {0}`,
  [CompileErrorCodes.UNBALANCED_CLOSING_BRACE]: `Unbalanced closing brace`,
  [CompileErrorCodes.UNTERMINATED_CLOSING_BRACE]: `Unterminated closing brace`,
  [CompileErrorCodes.EMPTY_PLACEHOLDER]: `Empty placeholder`,
  [CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER]: `Not allowed nest placeholder`,
  [CompileErrorCodes.INVALID_LINKED_FORMAT]: `Invalid linked format`,
  // parser error messages
  [CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL]: `Plural must have messages`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER]: `Unexpected empty linked modifier`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY]: `Unexpected empty linked key`,
  [CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS]: `Unexpected lexical analysis in token: '{0}'`
}

export function createCompileError<T extends number>(
  code: T,
  loc: SourceLocation | null,
  options: CreateCompileErrorOptions = {}
): CompileError {
  const { domain, messages, args } = options
  const msg = __DEV__
    ? format((messages || errorMessages)[code] || '', ...(args || []))
    : code
  const error = new SyntaxError(String(msg)) as CompileError
  error.code = code
  if (loc) {
    error.location = loc
  }
  error.domain = domain
  return error
}

/** @internal */
export function defaultOnError(error: CompileError): never {
  throw error
}
