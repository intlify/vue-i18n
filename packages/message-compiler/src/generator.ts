import { isString } from '@intlify/shared'
import { SourceMapGenerator, RawSourceMap } from 'source-map'
import {
  ResourceNode,
  Node,
  NodeTypes,
  PluralNode,
  MessageNode,
  TextNode,
  ListNode,
  NamedNode,
  LinkedNode,
  LinkedKeyNode,
  LinkedModitierNode,
  LiteralNode
} from './parser'
import { Position, LocationStub } from './location'
import { CodeGenOptions } from './options'
import { HelperNameMap } from './helpers'

export interface CodeGenResult {
  code: string
  ast: ResourceNode
  map?: RawSourceMap
}

type CodeGenContext = {
  source: string
  code: string
  indentLevel: number
  filename: string
  line: number
  breakLineCode: string
  needIndent: boolean
  column: number
  offset: number
  map?: SourceMapGenerator
}

type CodeGenNode =
  | TextNode
  | NamedNode
  | ListNode
  | LiteralNode
  | LinkedKeyNode
  | LinkedModitierNode

type CodeGenerator = {
  context(): CodeGenContext
  push(code: string, node?: CodeGenNode): void
  indent(withNewLine?: boolean): void
  deindent(withNewLine?: boolean): void
  newline(): void
  helper(key: string): string
  needIndent(): boolean
}

function createCodeGenerator(
  ast: ResourceNode,
  options: CodeGenOptions
): CodeGenerator {
  const {
    sourceMap,
    filename,
    breakLineCode,
    needIndent: _needIndent
  } = options
  const _context = {
    source: ast.loc!.source,
    filename,
    code: '',
    column: 1,
    line: 1,
    offset: 0,
    map: undefined,
    breakLineCode,
    needIndent: _needIndent,
    indentLevel: 0
  } as CodeGenContext

  const context = (): CodeGenContext => _context

  function push(code: string, node?: CodeGenNode): void {
    _context.code += code
    if (!__BROWSER__ && _context.map) {
      if (node && node.loc && node.loc !== LocationStub) {
        addMapping(node.loc.start, getMappingName(node))
      }
      advancePositionWithSource(_context, code)
    }
  }

  function _newline(n: number, withBreakLine = true): void {
    const _breakLineCode = withBreakLine ? breakLineCode : ''
    push(_needIndent ? _breakLineCode! + `  `.repeat(n) : _breakLineCode!)
  }

  function indent(withNewLine = true): void {
    const level = ++_context.indentLevel
    withNewLine && _newline(level)
  }

  function deindent(withNewLine = true): void {
    const level = --_context.indentLevel
    withNewLine && _newline(level)
  }

  function newline(): void {
    _newline(_context.indentLevel)
  }

  const helper = (key: string): string => `_${key}`
  const needIndent = (): boolean => _context.needIndent

  function addMapping(loc: Position, name?: string) {
    _context.map!.addMapping({
      name,
      source: _context.filename,
      original: {
        line: loc.line,
        column: loc.column - 1
      },
      generated: {
        line: _context.line,
        column: _context.column - 1
      }
    })
  }

  if (!__BROWSER__ && sourceMap) {
    _context.map = new SourceMapGenerator()
    _context.map!.setSourceContent(filename!, _context.source)
  }

  return {
    context,
    push,
    indent,
    deindent,
    newline,
    helper,
    needIndent
  }
}

function generateLinkedNode(generator: CodeGenerator, node: LinkedNode): void {
  const { helper } = generator
  generator.push(`${helper(HelperNameMap.LINKED)}(`)
  generateNode(generator, node.key)
  if (node.modifier) {
    generator.push(`, `)
    generateNode(generator, node.modifier)
  }
  generator.push(`)`)
}

function generateMessageNode(
  generator: CodeGenerator,
  node: MessageNode
): void {
  const { helper, needIndent } = generator
  generator.push(`${helper(HelperNameMap.NORMALIZE)}([`)
  generator.indent(needIndent())
  const length = node.items.length
  for (let i = 0; i < length; i++) {
    generateNode(generator, node.items[i])
    if (i === length - 1) {
      break
    }
    generator.push(', ')
  }
  generator.deindent(needIndent())
  generator.push('])')
}

function generatePluralNode(generator: CodeGenerator, node: PluralNode): void {
  const { helper, needIndent } = generator
  if (node.cases.length > 1) {
    generator.push(`${helper(HelperNameMap.PLURAL)}([`)
    generator.indent(needIndent())
    const length = node.cases.length
    for (let i = 0; i < length; i++) {
      generateNode(generator, node.cases[i])
      if (i === length - 1) {
        break
      }
      generator.push(', ')
    }
    generator.deindent(needIndent())
    generator.push(`])`)
  }
}

function generateResource(generator: CodeGenerator, node: ResourceNode): void {
  if (node.body) {
    generateNode(generator, node.body)
  } else {
    generator.push('null')
  }
}

function generateNode(generator: CodeGenerator, node: Node): void {
  const { helper } = generator
  switch (node.type) {
    case NodeTypes.Resource:
      generateResource(generator, node as ResourceNode)
      break
    case NodeTypes.Plural:
      generatePluralNode(generator, node as PluralNode)
      break
    case NodeTypes.Message:
      generateMessageNode(generator, node as MessageNode)
      break
    case NodeTypes.Linked:
      generateLinkedNode(generator, node as LinkedNode)
      break
    case NodeTypes.LinkedModifier:
      generator.push(
        JSON.stringify((node as LinkedModitierNode).value),
        node as LinkedModitierNode
      )
      break
    case NodeTypes.LinkedKey:
      generator.push(
        JSON.stringify((node as LinkedKeyNode).value),
        node as LinkedKeyNode
      )
      break
    case NodeTypes.List:
      generator.push(
        `${helper(HelperNameMap.INTERPOLATE)}(${helper(HelperNameMap.LIST)}(${
          (node as ListNode).index
        }))`,
        node as ListNode
      )
      break
    case NodeTypes.Named:
      generator.push(
        `${helper(HelperNameMap.INTERPOLATE)}(${helper(
          HelperNameMap.NAMED
        )}(${JSON.stringify((node as NamedNode).key)}))`,
        node as NamedNode
      )
      break
    case NodeTypes.Literal:
      generator.push(
        JSON.stringify((node as LiteralNode).value),
        node as LiteralNode
      )
      break
    case NodeTypes.Text:
      generator.push(JSON.stringify((node as TextNode).value), node as TextNode)
      break
    default:
      if (__DEV__) {
        throw new Error(`unhandled codegen node type: ${node.type}`)
      }
  }
}

// generate code from AST
export const generate = (
  ast: ResourceNode,
  options: CodeGenOptions = {} // eslint-disable-line
): CodeGenResult => {
  const mode = isString(options.mode) ? options.mode : 'normal'
  const filename = isString(options.filename)
    ? options.filename
    : 'message.intl'
  const sourceMap = !!options.sourceMap
  // prettier-ignore
  const breakLineCode = options.breakLineCode != null
    ? options.breakLineCode
    : mode === 'arrow'
      ? ';'
      : '\n'
  const needIndent = options.needIndent ? options.needIndent : mode !== 'arrow'
  const helpers = ast.helpers || []
  const generator = createCodeGenerator(ast, {
    mode,
    filename,
    sourceMap,
    breakLineCode,
    needIndent
  })

  generator.push(mode === 'normal' ? `function __msg__ (ctx) {` : `(ctx) => {`)
  generator.indent(needIndent)

  if (helpers.length > 0) {
    generator.push(
      `const { ${helpers.map(s => `${s}: _${s}`).join(', ')} } = ctx`
    )
    generator.newline()
  }

  generator.push(`return `)
  generateNode(generator, ast)
  generator.deindent(needIndent)
  generator.push(`}`)

  const { code, map } = generator.context()
  return {
    ast,
    code,
    map: map ? (map as any).toJSON() : undefined // eslint-disable-line @typescript-eslint/no-explicit-any
  }
}

function getMappingName(node: CodeGenNode): string | undefined {
  switch (node.type) {
    case NodeTypes.Text:
      return node.value
    case NodeTypes.List:
      return node.index.toString()
    case NodeTypes.Named:
      return node.key
    case NodeTypes.Literal:
      return node.value
    case NodeTypes.LinkedModifier:
      return node.value
    case NodeTypes.LinkedKey:
      return node.value
    default:
      return undefined
  }
}

function advancePositionWithSource(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  let linesCount = 0
  let lastNewLinePos = -1
  for (let i = 0; i < numberOfCharacters; i++) {
    if (source.charCodeAt(i) === 10 /* newline char code */) {
      linesCount++
      lastNewLinePos = i
    }
  }

  pos.offset += numberOfCharacters
  pos.line += linesCount
  pos.column =
    lastNewLinePos === -1
      ? pos.column + numberOfCharacters
      : numberOfCharacters - lastNewLinePos

  return pos
}
