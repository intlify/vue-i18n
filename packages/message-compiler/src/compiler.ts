import { CompileOptions } from './options'
import { createParser } from './parser'
import { transform } from './transformer'
import { generate, CodeGenResult } from './generator'
import { assign } from '@intlify/shared'

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CodeGenResult {
  const assignedOptions = assign({}, options)

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, assignedOptions)

  // generate javascript codes
  return generate(ast, assignedOptions)
}
