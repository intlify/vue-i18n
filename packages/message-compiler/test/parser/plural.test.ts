import {
  createParser,
  NodeTypes,
  PluralNode,
  ERROR_DOMAIN
} from '../../src/parser'
import { CompileErrorCodes, errorMessages } from '../../src/errors'

let spy: any // eslint-disable-line @typescript-eslint/no-explicit-any
beforeEach(() => {
  spy = jest.fn()
})

describe('basic', () => {
  test(`no apples | one apple  |  too much apples `, () => {
    const text = 'no apples | one apple  |  too much apples '
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Plural)
    const plural = ast.body as PluralNode
    expect(plural.cases).toHaveLength(3)
    expect(plural.cases).toMatchObject([
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'no apples'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'one apple'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'too much apples '
          }
        ]
      }
    ])
  })
})

describe('included new line', () => {
  test(`no apples |\n one apple  |\n  too much apples  `, () => {
    const text = 'no apples |\n one apple  |\n  too much apples  '
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Plural)
    const plural = ast.body as PluralNode
    expect(plural.cases).toHaveLength(3)
    expect(plural.cases).toMatchObject([
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'no apples'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'one apple'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'too much apples  '
          }
        ]
      }
    ])
  })
})

describe('complex usage', () => {
  // eslint-disable-next-line no-irregular-whitespace
  test(`@.lower:{'no apples'} | {1} apple | {count}　apples`, () => {
    // eslint-disable-next-line no-irregular-whitespace
    const text = `@.lower:{'no apples'} | {1} apple | {count}　apples`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Plural)
    const plural = ast.body as PluralNode
    expect(plural.cases).toHaveLength(3)
    expect(plural.cases).toMatchObject([
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Linked,
            key: {
              type: NodeTypes.Literal,
              value: 'no apples'
            },
            modifier: {
              type: NodeTypes.LinkedModifier,
              value: 'lower'
            }
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.List,
            index: 1
          },
          {
            type: NodeTypes.Text,
            value: ' apple'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Named,
            key: 'count'
          },
          {
            type: NodeTypes.Text,
            value: '　apples'
          }
        ]
      }
    ])
  })
})

describe('empty message', () => {
  test(` | | |`, () => {
    const text = ` | | |`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).toHaveBeenCalled()
    expect(
      spy.mock.calls.map(([err]: Array<Error>) => ({
        ...err,
        message: err.message
      })) // @ts-ignore
    ).toEqual([
      {
        domain: ERROR_DOMAIN,
        code: CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL,
        message: errorMessages[CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL],
        location: {
          start: {
            line: 1,
            offset: 0,
            column: 1
          },
          end: {
            line: 1,
            offset: 6,
            column: 7
          }
        }
      }
    ])

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Plural)
    const plural = ast.body as PluralNode
    expect(plural.cases).toHaveLength(4)
    expect(plural.cases).toMatchObject([
      {
        type: NodeTypes.Message,
        items: []
      },
      {
        type: NodeTypes.Message,
        items: []
      },
      {
        type: NodeTypes.Message,
        items: []
      },
      {
        type: NodeTypes.Message,
        items: []
      }
    ])
  })
})

describe('one empty message', () => {
  test(` foo | | bar | buz`, () => {
    const text = ` foo | | bar | buz`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).toHaveBeenCalled()
    expect(
      spy.mock.calls.map(([err]: Array<Error>) => ({
        ...err,
        message: err.message
      }))
    ).toEqual([
      {
        domain: ERROR_DOMAIN,
        code: CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL,
        message: errorMessages[CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL],
        location: {
          start: {
            line: 1,
            offset: 0,
            column: 1
          },
          end: {
            line: 1,
            offset: 18,
            column: 19
          }
        }
      }
    ])

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Plural)
    const plural = ast.body as PluralNode
    expect(plural.cases).toHaveLength(4)
    expect(plural.cases).toMatchObject([
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: ' foo'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: []
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'bar'
          }
        ]
      },
      {
        type: NodeTypes.Message,
        items: [
          {
            type: NodeTypes.Text,
            value: 'buz'
          }
        ]
      }
    ])
  })
})

test.todo(`@.lower: {'no apples'} | {1 apple | @:{count　apples`) // eslint-disable-line no-irregular-whitespace
