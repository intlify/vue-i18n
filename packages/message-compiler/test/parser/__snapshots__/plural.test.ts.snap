// Vitest Snapshot v1

exports[`basic > no apples | one apple  |  too much apples  1`] = `
{
  "body": {
    "cases": [
      {
        "end": 0,
        "items": [
          {
            "end": 9,
            "loc": {
              "end": {
                "column": 10,
                "line": 1,
                "offset": 9,
              },
              "start": {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "start": 0,
            "type": 3,
            "value": "no apples",
          },
        ],
        "loc": {
          "end": {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      {
        "end": 12,
        "items": [
          {
            "end": 21,
            "loc": {
              "end": {
                "column": 22,
                "line": 1,
                "offset": 21,
              },
              "start": {
                "column": 13,
                "line": 1,
                "offset": 12,
              },
            },
            "start": 12,
            "type": 3,
            "value": "one apple",
          },
        ],
        "loc": {
          "end": {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 2,
      },
      {
        "end": 42,
        "items": [
          {
            "end": 42,
            "loc": {
              "end": {
                "column": 43,
                "line": 1,
                "offset": 42,
              },
              "start": {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
            },
            "start": 26,
            "type": 3,
            "value": "too much apples ",
          },
        ],
        "loc": {
          "end": {
            "column": 43,
            "line": 1,
            "offset": 42,
          },
          "start": {
            "column": 27,
            "line": 1,
            "offset": 26,
          },
        },
        "start": 26,
        "type": 2,
      },
    ],
    "end": 42,
    "loc": {
      "end": {
        "column": 43,
        "line": 1,
        "offset": 42,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 42,
  "loc": {
    "end": {
      "column": 43,
      "line": 1,
      "offset": 42,
    },
    "source": "no apples | one apple  |  too much apples ",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`complex usage > @.lower:{'no apples'} | {1} apple | {count}　apples 1`] = `
{
  "body": {
    "cases": [
      {
        "end": 20,
        "items": [
          {
            "end": 21,
            "key": {
              "end": 21,
              "loc": {
                "end": {
                  "column": 22,
                  "line": 1,
                  "offset": 21,
                },
                "start": {
                  "column": 9,
                  "line": 1,
                  "offset": 8,
                },
              },
              "start": 8,
              "type": 9,
              "value": "no apples",
            },
            "loc": {
              "end": {
                "column": 22,
                "line": 1,
                "offset": 21,
              },
              "start": {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "modifier": {
              "end": 7,
              "loc": {
                "end": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "start": 1,
              "type": 8,
              "value": "lower",
            },
            "start": 0,
            "type": 6,
          },
        ],
        "loc": {
          "end": {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      {
        "end": 27,
        "items": [
          {
            "end": 27,
            "index": 1,
            "loc": {
              "end": {
                "column": 28,
                "line": 1,
                "offset": 27,
              },
              "start": {
                "column": 25,
                "line": 1,
                "offset": 24,
              },
            },
            "start": 24,
            "type": 5,
          },
          {
            "end": 33,
            "loc": {
              "end": {
                "column": 34,
                "line": 1,
                "offset": 33,
              },
              "start": {
                "column": 28,
                "line": 1,
                "offset": 27,
              },
            },
            "start": 27,
            "type": 3,
            "value": " apple",
          },
        ],
        "loc": {
          "end": {
            "column": 34,
            "line": 1,
            "offset": 33,
          },
          "start": {
            "column": 25,
            "line": 1,
            "offset": 24,
          },
        },
        "start": 24,
        "type": 2,
      },
      {
        "end": 50,
        "items": [
          {
            "end": 43,
            "key": "count",
            "loc": {
              "end": {
                "column": 44,
                "line": 1,
                "offset": 43,
              },
              "start": {
                "column": 37,
                "line": 1,
                "offset": 36,
              },
            },
            "start": 36,
            "type": 4,
          },
          {
            "end": 50,
            "loc": {
              "end": {
                "column": 51,
                "line": 1,
                "offset": 50,
              },
              "start": {
                "column": 44,
                "line": 1,
                "offset": 43,
              },
            },
            "start": 43,
            "type": 3,
            "value": "　apples",
          },
        ],
        "loc": {
          "end": {
            "column": 51,
            "line": 1,
            "offset": 50,
          },
          "start": {
            "column": 37,
            "line": 1,
            "offset": 36,
          },
        },
        "start": 36,
        "type": 2,
      },
    ],
    "end": 50,
    "loc": {
      "end": {
        "column": 51,
        "line": 1,
        "offset": 50,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 50,
  "loc": {
    "end": {
      "column": 51,
      "line": 1,
      "offset": 50,
    },
    "source": "@.lower:{'no apples'} | {1} apple | {count}　apples",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`empty message >  | | | 1`] = `
{
  "body": {
    "cases": [
      {
        "end": 0,
        "items": [],
        "loc": {
          "end": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      {
        "end": 0,
        "items": [],
        "loc": {
          "end": {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 2,
      },
      {
        "end": 3,
        "items": [],
        "loc": {
          "end": {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
          "start": {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
        },
        "start": 5,
        "type": 2,
      },
      {
        "end": 6,
        "items": [],
        "loc": {
          "end": {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 2,
      },
    ],
    "end": 6,
    "loc": {
      "end": {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 6,
  "loc": {
    "end": {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "source": " | | |",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`included new line > no apples |
 one apple  |
  too much apples   1`] = `
{
  "body": {
    "cases": [
      {
        "end": 0,
        "items": [
          {
            "end": 9,
            "loc": {
              "end": {
                "column": 10,
                "line": 1,
                "offset": 9,
              },
              "start": {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "start": 0,
            "type": 3,
            "value": "no apples",
          },
        ],
        "loc": {
          "end": {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      {
        "end": 13,
        "items": [
          {
            "end": 22,
            "loc": {
              "end": {
                "column": 11,
                "line": 2,
                "offset": 22,
              },
              "start": {
                "column": 2,
                "line": 2,
                "offset": 13,
              },
            },
            "start": 13,
            "type": 3,
            "value": "one apple",
          },
        ],
        "loc": {
          "end": {
            "column": 11,
            "line": 2,
            "offset": 22,
          },
          "start": {
            "column": 2,
            "line": 2,
            "offset": 13,
          },
        },
        "start": 13,
        "type": 2,
      },
      {
        "end": 45,
        "items": [
          {
            "end": 45,
            "loc": {
              "end": {
                "column": 20,
                "line": 3,
                "offset": 45,
              },
              "start": {
                "column": 3,
                "line": 3,
                "offset": 28,
              },
            },
            "start": 28,
            "type": 3,
            "value": "too much apples  ",
          },
        ],
        "loc": {
          "end": {
            "column": 20,
            "line": 3,
            "offset": 45,
          },
          "start": {
            "column": 3,
            "line": 3,
            "offset": 28,
          },
        },
        "start": 28,
        "type": 2,
      },
    ],
    "end": 45,
    "loc": {
      "end": {
        "column": 20,
        "line": 3,
        "offset": 45,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 45,
  "loc": {
    "end": {
      "column": 20,
      "line": 3,
      "offset": 45,
    },
    "source": "no apples |
 one apple  |
  too much apples  ",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`one empty message >  foo | | bar | buz 1`] = `
{
  "body": {
    "cases": [
      {
        "end": 0,
        "items": [
          {
            "end": 4,
            "loc": {
              "end": {
                "column": 5,
                "line": 1,
                "offset": 4,
              },
              "start": {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "start": 0,
            "type": 3,
            "value": " foo",
          },
        ],
        "loc": {
          "end": {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      {
        "end": 4,
        "items": [],
        "loc": {
          "end": {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
          "start": {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
        },
        "start": 7,
        "type": 2,
      },
      {
        "end": 9,
        "items": [
          {
            "end": 12,
            "loc": {
              "end": {
                "column": 13,
                "line": 1,
                "offset": 12,
              },
              "start": {
                "column": 10,
                "line": 1,
                "offset": 9,
              },
            },
            "start": 9,
            "type": 3,
            "value": "bar",
          },
        ],
        "loc": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "start": 9,
        "type": 2,
      },
      {
        "end": 18,
        "items": [
          {
            "end": 18,
            "loc": {
              "end": {
                "column": 19,
                "line": 1,
                "offset": 18,
              },
              "start": {
                "column": 16,
                "line": 1,
                "offset": 15,
              },
            },
            "start": 15,
            "type": 3,
            "value": "buz",
          },
        ],
        "loc": {
          "end": {
            "column": 19,
            "line": 1,
            "offset": 18,
          },
          "start": {
            "column": 16,
            "line": 1,
            "offset": 15,
          },
        },
        "start": 15,
        "type": 2,
      },
    ],
    "end": 18,
    "loc": {
      "end": {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 18,
  "loc": {
    "end": {
      "column": 19,
      "line": 1,
      "offset": 18,
    },
    "source": " foo | | bar | buz",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
